File: src/us/achromaticmetaphor/imcktg/TTS.java
package us.achromaticmetaphor.imcktg;

import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Semaphore;

import android.speech.tts.TextToSpeech;

public class TTS extends ToneGenerator implements TextToSpeech.OnUtteranceCompletedListener {

  private TextToSpeech tts;
  private Map<String, Semaphore> semas;

  public TTS(TextToSpeech tts) {
    this(tts, 1.0f, 0.8f);
  }

  public TTS(TextToSpeech tts, float pitch, float srate) {
    this.tts = tts;
    tts.setOnUtteranceCompletedListener(this);
    tts.setPitch(pitch);
    tts.setSpeechRate(srate);
    semas = new ConcurrentHashMap<String, Semaphore>();
  }

  @Override
  public void writeTone(File tone, String s, boolean extend) throws IOException {
    String uid = tone.getAbsolutePath();
    semas.put(uid, new Semaphore(0));
    HashMap<String, String> params = new HashMap<String, String>();
    params.put(TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID, uid);
    tts.synthesizeToFile(s, params, Tone.tmpfile(tone).getAbsolutePath());
    semas.get(uid).acquireUninterruptibly();
    semas.remove(uid);
    Tone.tmpRename(tone);
    if (extend)
      Tone.waveAppendSilence(tone, 2);
  }

  @Override
  public String filenameExt() {
    return ".wav";
  }

  @Override
  public void writeTone(OutputStream out, String s, boolean extend) throws IOException {
    throw new IllegalArgumentException("method not implemented");
  }

  @Override
  public String filenameTypePrefix() {
    return "TextToSpeech:";
  }

  @Override
  public void onUtteranceCompleted(String uid) {
    semas.get(uid).release();
  }

}

File: src/main/java/us/achromaticmetaphor/imcktg/TTS.kt
package us.achromaticmetaphor.imcktg

import android.content.Context
import android.speech.tts.TextToSpeech
import android.speech.tts.TextToSpeech.OnUtteranceCompletedListener
import us.achromaticmetaphor.imcktg.Tone.Companion.tmpRename
import us.achromaticmetaphor.imcktg.Tone.Companion.tmpfile
import java.io.File
import java.io.IOException
import java.io.OutputStream
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Semaphore

class TTS @JvmOverloads constructor(private val tts: TextToSpeech, pitch: Float = 1.0f, private val srate: Float = 0.8f, private val repeatCount: Int = 0, private val context: Context) : ToneGenerator(), OnUtteranceCompletedListener {
    private val semas = ConcurrentHashMap<String, Semaphore>()

    @Throws(IOException::class)
    override fun writeTone(out: File, s: String, extend: Boolean) {
        val uid = UUID.randomUUID().toString()
        semas[uid] = Semaphore(0)
        val params = HashMap<String, String>()
        params[TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID] = uid
        tts.synthesizeToFile(s, params, tmpfile(out).absolutePath)
        semas[uid]!!.acquireUninterruptibly()
        semas.remove(uid)
        tmpRename(out)
        WAVETone(out).apply{
            if (extend) appendSilence((2000 / srate).toInt())
            if (repeatCount > 0) repeat(repeatCount)
            close()
        }
    }

    @Throws(IOException::class)
    override fun writeTone(out: OutputStream, s: String) {
        val tmpfile = File(context.cacheDir, "IMCKTG.speech")
        writeTone(tmpfile, s, false)
        tmpfile.inputStream().use { input ->
            input.copyTo(out)
        }
    }
    override fun filenameTypePrefix() = "TextToSpeech:$repeatCount:"
    override fun onUtteranceCompleted(uid: String) = semas[uid]!!.release()
    override fun filenameExt() = ".wav"

    init {
        tts.setOnUtteranceCompletedListener(this)
        tts.setPitch(pitch)
        tts.setSpeechRate(srate)
    }
}